File: mms_web/bot_schedules.py
"""Periodic Bot schedules: durable entities independent from a task.

A schedule is a small rule (``once`` / ``interval`` / ``daily`` / ``weekly``)
plus the prompt to run.  Every calculation here is a pure function over plain
data so the scheduler thread in :mod:`mms_web.bots` keeps only thin wiring and
the timing rules stay unit-testable.  No cron expressions: users say "every
three hours" or "every day at 9", and those four shapes cover that.

Timezone arithmetic uses the standard library ``zoneinfo`` only.  A schedule
always stores its IANA name, otherwise "daily 09:00" drifts across a daylight
saving change.
"""
from __future__ import annotations

import os
import re
import time
from copy import deepcopy
from datetime import datetime, timedelta, timezone
from uuid import uuid4
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from .errors import WebError

MIN_INTERVAL_SECONDS = 300
MAX_SCHEDULES_PER_BOT = 20
RECENT_TASK_LIMIT = 10
MAX_PROMPT_CHARS = 32000
KINDS = ("once", "interval", "daily", "weekly")
OVERLAP_POLICIES = ("skip", "queue")
CREATORS = ("user", "bot")

_LOCAL_TIME_RE = re.compile(r"^([01]\d|2[0-3]):([0-5]\d)$")
_EVERY_RE = re.compile(r"^(\d{1,9})([hms]?)$")
_FIXED_OFFSET_RE = re.compile(r"^UTC([+-])(\d{2}):(\d{2})$")
_ZONEINFO_ROOTS = ("/var/db/timezone/zoneinfo", "/usr/share/zoneinfo", "/usr/lib/zoneinfo", "/etc/zoneinfo")
_WEEKDAYS = {"mon": 0, "tue": 1, "wed": 2, "thu": 3, "fri": 4, "sat": 5, "sun": 6,
             "周一": 0, "周二": 1, "周三": 2, "周四": 3, "周五": 4, "周六": 5, "周日": 6}


# --------------------------------------------------------------------- helpers
def _parse_iso(value, code: str, message: str) -> datetime:
    try:
        parsed = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
        if parsed.tzinfo is None:
            raise ValueError()
        return parsed.astimezone(timezone.utc)
    except (TypeError, ValueError):
        raise WebError(code, message, 400) from None


def _iso(value: datetime | None) -> str | None:
    return value.astimezone(timezone.utc).isoformat() if value else None


def _zone(name: str):
    fixed = _FIXED_OFFSET_RE.fullmatch(str(name or ""))
    if fixed:
        sign = 1 if fixed.group(1) == "+" else -1
        delta = timedelta(hours=int(fixed.group(2)), minutes=int(fixed.group(3)))
        return timezone(sign * delta)
    return ZoneInfo(name)


def _iana_from_path(path: str) -> str | None:
    for root in _ZONEINFO_ROOTS:
        if path.startswith(root + "/"):
            name = path[len(root) + 1:]
            try:
                ZoneInfo(name)
                return name
            except (ZoneInfoNotFoundError, ValueError):
                return None
    return None


def local_timezone_name() -> str:
    """The machine's IANA timezone, or a fixed ``UTC±HH:MM`` fallback.

    There is no stdlib API for "the local IANA name", so try the environment,
    the ``/etc/localtime`` symlink and the platform name before falling back to
    a fixed offset.  The result is always written back into the schedule, so a
    later read on another machine still sees an explicit timezone.
    """
    candidate = os.environ.get("TZ", "").strip()
    if candidate:
        try:
            _zone(candidate)
            return candidate
        except (ZoneInfoNotFoundError, ValueError):
            pass
    try:
        resolved = str(os.path.realpath("/etc/localtime"))
    except OSError:
        resolved = ""
    if resolved and resolved != "/etc/localtime":
        name = _iana_from_path(resolved)
        if name:
            return name
    platform_name = time.tzname[0] if time.tzname else ""
    if platform_name:
        try:
            _zone(platform_name)
            return platform_name
        except (ZoneInfoNotFoundError, ValueError):
            pass
    offset = datetime.now().astimezone().utcoffset() or timedelta(0)
    total = int(offset.total_seconds())
    sign = "+" if total >= 0 else "-"
    total = abs(total)
    return f"UTC{sign}{total // 3600:02d}:{total % 3600 // 60:02d}"


# --------------------------------------------------------------- rule parsing
def normalize_timezone(value: str | None) -> str:
    """Validate an IANA name, defaulting to (and always storing) the machine one."""
    if value in (None, ""):
        return local_timezone_name()
    if not isinstance(value, str) or len(value) > 100:
        raise WebError("INVALID_TIMEZONE", "时区必须是 IANA 名称，例如 Asia/Singapore。", 400)
    name = value.strip()
    if not name:
        return local_timezone_name()
    try:
        _zone(name)
    except (ZoneInfoNotFoundError, ValueError):
        raise WebError("INVALID_TIMEZONE",
                       f"无法识别时区 {name[:40]}，请使用 IANA 名称，例如 Asia/Singapore。", 400) from None
    return name


def parse_every(value) -> int:
    """Parse ``3h`` / ``180m`` / ``10800s`` / a bare second count."""
    match = _EVERY_RE.fullmatch(str(value or "").strip().lower())
    if not match:
        raise WebError("INVALID_SCHEDULE_RULE", "间隔要写成 3h、180m、10800s 或直接的秒数。", 400)
    seconds = int(match.group(1)) * {"h": 3600, "m": 60, "s": 1}[match.group(2) or "s"]
    if seconds <= 0:
        raise WebError("INVALID_SCHEDULE_RULE", "间隔必须大于 0 秒。", 400)
    return seconds


def parse_weekday(value) -> int:
    """Parse ``mon``…``sun`` (also ``周一``…``周日``) into 0=Monday…6=Sunday."""
    text = str(value or "").strip().lower()
    if text in _WEEKDAYS:
        return _WEEKDAYS[text]
    raise WebError("INVALID_SCHEDULE_RULE", "weekday 要写成 mon…sun（0=周一，6=周日）。", 400)


def normalize_rule(rule: dict) -> dict:
    """Validate and normalize one of the four supported rule shapes.

    An unknown ``kind`` is an error, never silently treated as ``once``; adding
    a ``cron`` kind later means extending this function and nothing else.
    """
    if not isinstance(rule, dict):
        raise WebError("INVALID_SCHEDULE_RULE", "定时规则必须是对象。", 400)
    kind = rule.get("kind")
    if kind == "once":
        at = _parse_iso(rule.get("at"), "INVALID_SCHEDULE_RULE", "once 定时需要包含时区的 ISO8601 时间。")
        return {"kind": "once", "at": _iso(at)}
    if kind == "interval":
        every = rule.get("everySeconds")
        if type(every) is not int or every <= 0:
            raise WebError("INVALID_SCHEDULE_RULE", "everySeconds 必须是正整数秒。", 400)
        if every < MIN_INTERVAL_SECONDS:
            raise WebError("SCHEDULE_INTERVAL_TOO_SHORT", "定时间隔最短 5 分钟。", 400)
        return {"kind": "interval", "everySeconds": every}
    if kind in {"daily", "weekly"}:
        at_local = rule.get("atLocalTime")
        local = at_local.strip() if isinstance(at_local, str) else ""
        if not _LOCAL_TIME_RE.fullmatch(local):
            raise WebError("INVALID_SCHEDULE_RULE", "atLocalTime 必须是 HH:MM（24 小时制）。", 400)
        normalized = {"kind": kind, "atLocalTime": local}
        if kind == "weekly":
            weekday = rule.get("weekday")
            if type(weekday) is not int or not 0 <= weekday <= 6:
                raise WebError("INVALID_SCHEDULE_RULE", "weekday 必须是 0（周一）到 6（周日）的整数。", 400)
            normalized["weekday"] = weekday
        return normalized
    raise WebError("INVALID_SCHEDULE_RULE",
                   f"不支持的定时类型 {str(kind)[:40] or '（空）'}；支持 once、interval、daily、weekly。", 400)


# ----------------------------------------------------------------- next times
def _local_instant(wall: datetime, zone) -> datetime:
    """Resolve a naive local wall time to UTC, DST-aware.

    A nonexistent local time (spring forward) resolves to the transition
    instant, i.e. the first valid local time after the gap.  An ambiguous local
    time (fall back) resolves to its first occurrence (``fold=0``).
    """
    candidate = wall.replace(tzinfo=zone)
    first = candidate.astimezone(timezone.utc)
    if first.astimezone(zone).replace(tzinfo=None) == wall:
        return first
    before = wall.replace(tzinfo=zone, fold=0).utcoffset()
    after = wall.replace(tzinfo=zone, fold=1).utcoffset()
    if before == after:
        return first
    low = min(wall.replace(tzinfo=zone, fold=0).astimezone(timezone.utc),
              wall.replace(tzinfo=zone, fold=1).astimezone(timezone.utc))
    high = max(wall.replace(tzinfo=zone, fold=0).astimezone(timezone.utc),
               wall.replace(tzinfo=zone, fold=1).astimezone(timezone.utc))
    while high - low > timedelta(seconds=1):
        middle = low + (high - low) / 2
        if middle.astimezone(zone).utcoffset() == after:
            high = middle
        else:
            low = middle
    return high.replace(microsecond=0)


def next_run_at(rule: dict, timezone_name: str, *, after: datetime,
                previous: datetime | None = None) -> datetime | None:
    """First run strictly after ``after``; ``None`` only for a spent ``once``.

    ``previous`` is the nominal time that was just handled, not the wall clock
    moment it actually ran.  ``interval`` advances from it, so a late tick never
    accumulates drift.
    """
    kind = rule["kind"]
    if kind == "once":
        at = _parse_iso(rule["at"], "INVALID_SCHEDULE_RULE", "once 定时需要包含时区的 ISO8601 时间。")
        return at if at > after else None
    if kind == "interval":
        every = int(rule["everySeconds"])
        base = previous or after
        if base <= after:
            steps = int((after - base).total_seconds() // every) + 1
            base = base + timedelta(seconds=every * steps)
        return base
    if kind not in {"daily", "weekly"}:
        raise WebError("INVALID_SCHEDULE_RULE", "不支持的定时类型。", 400)
    zone = _zone(timezone_name)
    hour, minute = (int(part) for part in rule["atLocalTime"].split(":"))
    local = after.astimezone(zone)
    wall = local.replace(tzinfo=None, second=0, microsecond=0)
    candidate = wall.replace(hour=hour, minute=minute)
    if kind == "weekly":
        candidate += timedelta(days=(int(rule["weekday"]) - wall.weekday()) % 7)
    if candidate <= wall:
        candidate += timedelta(days=1 if kind == "daily" else 7)
    return _local_instant(candidate, zone)


def advance(schedule: dict, *, now: datetime) -> tuple[dict, int]:
    """Move one schedule past ``now`` and report how many runs were skipped.

    The only entry point for both catching up and moving on: a run missed by
    less than one cycle is fired now (the returned schedule keeps the original
    cadence); more than that is dropped and only counted, so an offline Pilot
    never replays a backlog.
    """
    updated = deepcopy(schedule)
    due = _parse_iso(updated.get("nextRunAt"), "INVALID_SCHEDULE_RULE", "nextRunAt 必须是包含时区的 ISO8601 时间。") \
        if updated.get("nextRunAt") else None
    if due is None or due > now:
        return updated, 0
    rule = updated["rule"]
    if rule["kind"] == "once":
        updated["nextRunAt"] = None
        return updated, 0
    if rule["kind"] == "interval":
        every = int(rule["everySeconds"])
        skipped = int((now - due).total_seconds() // every)
        updated["nextRunAt"] = _iso(due + timedelta(seconds=every * (skipped + 1)))
        return updated, skipped
    nxt = next_run_at(rule, updated["timezone"], after=due, previous=due)
    skipped = 0
    while nxt is not None and nxt <= now:
        skipped += 1
        nxt = next_run_at(rule, updated["timezone"], after=nxt, previous=nxt)
    updated["nextRunAt"] = _iso(nxt)
    return updated, skipped


# ------------------------------------------------------------------ lifecycle
def sanitize_schedules(raw) -> dict:
    """Keep stored rows usable without letting one bad row brick the loop.

    A row whose rule or timezone no longer resolves is kept but parked:
    disabled with no next run, so the UI can still show and fix it instead of
    silently losing a user's schedule.
    """
    if not isinstance(raw, dict):
        return {}
    clean = {}
    for key, row in raw.items():
        if not isinstance(row, dict) or not isinstance(row.get("botId"), str):
            continue
        if not isinstance(row.get("recentTaskIds"), list):
            row["recentTaskIds"] = []
        try:
            normalize_rule(row.get("rule"))
            normalize_timezone(row.get("timezone"))
            if row.get("nextRunAt"):
                _parse_iso(row["nextRunAt"], "INVALID_SCHEDULE_RULE", "")
        except (WebError, TypeError, ValueError):
            row.update(enabled=False, nextRunAt=None,
                       lastSkip={"at": row.get("updatedAt"), "reason": "invalid", "skipped": 0})
        clean[str(key)] = row
    return clean


def is_due(schedule: dict, now: datetime) -> bool:
    """Whether a stored ``nextRunAt`` is at or before ``now``."""
    due = str(schedule.get("nextRunAt") or "")
    if not due:
        return False
    try:
        return _parse_iso(due, "INVALID_SCHEDULE_RULE", "") <= now
    except WebError:
        return False


def validate_schedule(payload, existing_count: int) -> None:
    if not isinstance(payload, dict):
        raise WebError("INVALID_REQUEST", "定时内容必须是对象。", 400)
    if int(existing_count) >= MAX_SCHEDULES_PER_BOT:
        raise WebError("SCHEDULE_LIMIT", "一个 Bot 最多 20 条定时。", 409)
    prompt = payload.get("prompt")
    if not isinstance(prompt, str) or not prompt.strip() or len(prompt) > MAX_PROMPT_CHARS:
        raise WebError("INVALID_REQUEST", f"prompt 不能为空且长度须在 {MAX_PROMPT_CHARS} 字以内。", 400)
    if payload.get("overlapPolicy", "skip") not in OVERLAP_POLICIES:
        raise WebError("INVALID_REQUEST", "overlapPolicy 必须是 skip 或 queue。", 400)
    normalize_rule(payload.get("rule"))
    normalize_timezone(payload.get("timezone"))


def build_schedule(bot_id: str, payload: dict, *, existing_count: int, now: datetime,
                   created_by: str | None = None, schedule_id: str | None = None) -> dict:
    validate_schedule(payload, existing_count)
    rule = normalize_rule(payload.get("rule"))
    zone = normalize_timezone(payload.get("timezone"))
    creator = created_by or payload.get("createdBy") or "user"
    if creator not in CREATORS:
        raise WebError("INVALID_REQUEST", "createdBy 必须是 user 或 bot。", 400)
    stamp = now.astimezone(timezone.utc).isoformat(timespec="milliseconds")
    return {
        "id": schedule_id or ("sch_" + uuid4().hex[:16]),
        "botId": bot_id,
        "prompt": str(payload["prompt"]).strip(),
        "rule": rule,
        "timezone": zone,
        "enabled": True,
        "overlapPolicy": payload.get("overlapPolicy", "skip"),
        # A once rule that is already in the past still fires on the next tick.
        "nextRunAt": rule["at"] if rule["kind"] == "once" else _iso(next_run_at(rule, zone, after=now)),
        "lastRunAt": None,
        "lastTaskId": None,
        "recentTaskIds": [],
        "lastSkip": None,
        "createdBy": creator,
        "createdAt": stamp,
        "updatedAt": stamp,
    }


def apply_update(schedule: dict, payload: dict, *, now: datetime) -> dict:
    """Validate an edit (prompt / rule / timezone / overlapPolicy)."""
    if not isinstance(payload, dict):
        raise WebError("INVALID_REQUEST", "定时内容必须是对象。", 400)
    updated = deepcopy(schedule)
    if "prompt" in payload:
        prompt = payload.get("prompt")
        if not isinstance(prompt, str) or not prompt.strip() or len(prompt) > MAX_PROMPT_CHARS:
            raise WebError("INVALID_REQUEST", f"prompt 不能为空且长度须在 {MAX_PROMPT_CHARS} 字以内。", 400)
        updated["prompt"] = prompt.strip()
    if "overlapPolicy" in payload:
        if payload["overlapPolicy"] not in OVERLAP_POLICIES:
            raise WebError("INVALID_REQUEST", "overlapPolicy 必须是 skip 或 queue。", 400)
        updated["overlapPolicy"] = payload["overlapPolicy"]
    if "rule" in payload or "timezone" in payload:
        rule = normalize_rule(payload.get("rule") or updated["rule"])
        zone = normalize_timezone(payload.get("timezone", updated["timezone"]))
        updated["rule"], updated["timezone"] = rule, zone
        # An edited rule restarts from now; it never inherits a stale due time.
        updated["nextRunAt"] = rule["at"] if rule["kind"] == "once" else _iso(next_run_at(rule, zone, after=now))
    updated["updatedAt"] = now.astimezone(timezone.utc).isoformat(timespec="milliseconds")
    return updated

File: apps/mms-web/src/bot-schedules.ts
/** Schedule rule helpers and run-state labels for the Bot 定时 UI. */

export const MIN_INTERVAL_SECONDS = 300;
export const MAX_SCHEDULES_PER_BOT = 20;
export const MAX_PROMPT_CHARS = 32000;
export const WEEKDAY_LABELS = ["周一", "周二", "周三", "周四", "周五", "周六", "周日"] as const;
export const INTERVAL_HOUR_PRESETS = [1, 3, 6, 12] as const;

export type ScheduleKind = "once" | "interval" | "daily" | "weekly";
export type OverlapPolicy = "skip" | "queue";
export type SkipReason = "paused" | "busy" | "missed" | "error" | "invalid";

export type ScheduleRule =
  | { kind: "once"; at: string }
  | { kind: "interval"; everySeconds: number }
  | { kind: "daily"; atLocalTime: string }
  | { kind: "weekly"; weekday: number; atLocalTime: string };

export interface ScheduleLastSkip {
  at?: string;
  reason: SkipReason | string;
  skipped?: number;
}

export interface BotSchedule {
  id: string;
  botId: string;
  prompt: string;
  rule: ScheduleRule;
  timezone: string;
  enabled: boolean;
  overlapPolicy: OverlapPolicy | string;
  nextRunAt: string | null;
  lastRunAt: string | null;
  lastTaskId: string | null;
  recentTaskIds?: string[];
  lastSkip: ScheduleLastSkip | null;
  createdBy?: string;
  createdAt: string;
  updatedAt: string;
}

export interface ComposerScheduleForm {
  kind: ScheduleKind;
  onceDate: string;
  onceTime: string;
  atLocalTime: string;
  weekday: number;
  intervalHours: number;
}

export type ScheduleRunKind =
  | "upcoming"
  | "held"
  | "completed"
  | "invalid"
  | "error"
  | "disabled"
  | "gated"
  | "unknown";

export interface ScheduleRunState {
  kind: ScheduleRunKind;
  holdReason?: "paused" | "busy";
  label: string;
  detail: string;
}

function pad(value: number): string {
  return String(value).padStart(2, "0");
}

export function defaultComposerForm(now: Date = new Date()): ComposerScheduleForm {
  const tomorrow = new Date(now.getFullYear(), now.getMonth(), now.getDate() + 1);
  return {
    kind: "once",
    onceDate: `${tomorrow.getFullYear()}-${pad(tomorrow.getMonth() + 1)}-${pad(tomorrow.getDate())}`,
    onceTime: "09:00",
    atLocalTime: "09:00",
    weekday: 0,
    intervalHours: 3,
  };
}

export function intervalSecondsFromHours(hours: number): number {
  return Math.round(Number(hours) * 3600);
}

export function localTimezone(): string {
  try {
    return Intl.DateTimeFormat().resolvedOptions().timeZone || "UTC";
  } catch {
    return "UTC";
  }
}

export function padTime(value: string): string {
  const match = String(value || "").trim().match(/^(\d{1,2}):(\d{2})$/);
  if (!match) return "";
  const hours = Number(match[1]);
  const minutes = Number(match[2]);
  if (hours > 23 || minutes > 59) return "";
  return `${pad(hours)}:${pad(minutes)}`;
}

export function onceLocalToIso(date: string, time: string): string {
  const clock = padTime(time) || "09:00";
  const local = new Date(`${date}T${clock}:00`);
  if (Number.isNaN(local.valueOf())) {
    throw new Error("时间无效");
  }
  return local.toISOString();
}

export function composerRuleFromForm(form: ComposerScheduleForm): ScheduleRule {
  if (form.kind === "once") {
    return { kind: "once", at: onceLocalToIso(form.onceDate, form.onceTime) };
  }
  if (form.kind === "daily") {
    const atLocalTime = padTime(form.atLocalTime);
    if (!atLocalTime) throw new Error("时间无效");
    return { kind: "daily", atLocalTime };
  }
  if (form.kind === "weekly") {
    const atLocalTime = padTime(form.atLocalTime);
    if (!atLocalTime) throw new Error("时间无效");
    const weekday = Number(form.weekday);
    if (!Number.isInteger(weekday) || weekday < 0 || weekday > 6) {
      throw new Error("星期无效");
    }
    return { kind: "weekly", weekday, atLocalTime };
  }
  return {
    kind: "interval",
    everySeconds: intervalSecondsFromHours(form.intervalHours),
  };
}

export function validateComposerForm(form: ComposerScheduleForm): string | null {
  if (form.kind === "interval") {
    const seconds = intervalSecondsFromHours(form.intervalHours);
    if (!Number.isFinite(seconds) || seconds <= 0) return "请填写间隔小时数";
    if (seconds < MIN_INTERVAL_SECONDS) {
      return "定时间隔最短 5 分钟。太频繁会持续消耗模型额度。";
    }
    return null;
  }
  if (form.kind === "once") {
    if (!form.onceDate) return "请选择日期";
    if (!padTime(form.onceTime)) return "请填写 HH:MM 时间";
    try {
      onceLocalToIso(form.onceDate, form.onceTime);
    } catch {
      return "时间无效";
    }
    return null;
  }
  if (!padTime(form.atLocalTime)) return "请填写 HH:MM 时间";
  if (form.kind === "weekly") {
    const weekday = Number(form.weekday);
    if (!Number.isInteger(weekday) || weekday < 0 || weekday > 6) return "请选择星期";
  }
  return null;
}

export function describeInterval(seconds: number): string {
  if (!Number.isFinite(seconds) || seconds <= 0) return "周期间隔";
  if (seconds % 3600 === 0) {
    const hours = seconds / 3600;
    return `每 ${hours} 小时`;
  }
  if (seconds % 60 === 0) {
    const minutes = seconds / 60;
    return `每 ${minutes} 分钟`;
  }
  return `每 ${seconds} 秒`;
}

export function describeRule(rule?: ScheduleRule | null): string {
  if (!rule) return "";
  if (rule.kind === "once") {
    const date = new Date(rule.at);
    if (Number.isNaN(date.valueOf())) return "一次";
    return `一次 ${date.getMonth() + 1}月${date.getDate()}日 ${pad(date.getHours())}:${pad(date.getMinutes())}`;
  }
  if (rule.kind === "daily") return `每天 ${rule.atLocalTime}`;
  if (rule.kind === "weekly") {
    const label = WEEKDAY_LABELS[rule.weekday] || `周${rule.weekday}`;
    return `每周${label} ${rule.atLocalTime}`;
  }
  if (rule.kind === "interval") return describeInterval(rule.everySeconds);
  return "";
}

export function formFromRule(rule: ScheduleRule, now: Date = new Date()): ComposerScheduleForm {
  const form = defaultComposerForm(now);
  form.kind = rule.kind;
  if (rule.kind === "once") {
    const date = new Date(rule.at);
    if (!Number.isNaN(date.valueOf())) {
      form.onceDate = `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
      form.onceTime = `${pad(date.getHours())}:${pad(date.getMinutes())}`;
    }
    return form;
  }
  if (rule.kind === "daily") {
    form.atLocalTime = rule.atLocalTime;
    return form;
  }
  if (rule.kind === "weekly") {
    form.weekday = rule.weekday;
    form.atLocalTime = rule.atLocalTime;
    return form;
  }
  form.intervalHours = rule.everySeconds / 3600;
  return form;
}

export function describeComposerChip(form: ComposerScheduleForm): string {
  if (form.kind === "once") {
    if (!form.onceDate || !padTime(form.onceTime)) return "";
    return `${form.onceDate} ${padTime(form.onceTime)}`;
  }
  try {
    return describeRule(composerRuleFromForm(form));
  } catch {
    return "";
  }
}

function holdDetail(reason: string | undefined): string {
  return reason === "busy"
    ? "到点时上一轮还在跑，恢复后会补跑。"
    : "到点时被暂停或总闸拦住，恢复后会补跑。";
}

export function scheduleRunState(
  schedule: BotSchedule,
  now: Date = new Date(),
  options: { wakeEnabled?: boolean } = {},
): ScheduleRunState {
  const skip = schedule.lastSkip;
  const next = schedule.nextRunAt;
  const nextMs = next ? Date.parse(next) : Number.NaN;
  const hasNext = Number.isFinite(nextMs);
  const past = hasNext && nextMs <= now.getTime();
  const future = hasNext && nextMs > now.getTime();
  const holdSkip = skip?.reason === "paused" || skip?.reason === "busy";

  if (next == null && schedule.enabled === false && skip?.reason === "invalid") {
    return {
      kind: "invalid",
      label: "记录损坏",
      detail: "这条定时读不出来，改规则或时区后再启用。",
    };
  }
  if (next == null && schedule.lastRunAt) {
    return {
      kind: "completed",
      label: "已执行完",
      detail: "这一次已经跑过，不会再触发。",
    };
  }
  // Repeating rules keep a future nextRunAt while paused. That must not read as 待触发.
  if (!schedule.enabled) {
    if (holdSkip && schedule.lastRunAt == null) {
      return {
        kind: "disabled",
        holdReason: skip?.reason as "paused" | "busy",
        label: "已暂停",
        detail: holdDetail(skip?.reason),
      };
    }
    return {
      kind: "disabled",
      label: "已暂停",
      detail: "这条定时不会触发；恢复后按下次时间跑。",
    };
  }
  if (past && holdSkip && schedule.lastRunAt == null) {
    return {
      kind: "held",
      holdReason: skip?.reason as "paused" | "busy",
      label: "已到点但被挂起",
      detail: holdDetail(skip?.reason),
    };
  }
  if (skip?.reason === "error") {
    return {
      kind: "error",
      label: "上次没能建出任务",
      detail: "到点会再试，不是已经执行完。",
    };
  }
  if (options.wakeEnabled === false) {
    return {
      kind: "gated",
      label: "被总闸拦住",
      detail: "被自动唤醒总闸拦住",
    };
  }
  if (future) {
    return { kind: "upcoming", label: "待触发", detail: "" };
  }
  if (past) {
    return {
      kind: "upcoming",
      label: "待触发",
      detail: "已到点，正在等下一轮调度，不是卡住。",
    };
  }
  return { kind: "unknown", label: "", detail: "" };
}

export function skipNote(schedule: BotSchedule, now: Date = new Date()): string {
  const skip = schedule.lastSkip;
  if (!skip) return "";
  if (skip.reason === "missed") {
    const skipped = Number(skip.skipped) || 0;
    return skipped > 0
      ? `错过 ${skipped} 次（进程没开着，不补跑）`
      : "错过了周期（不补跑）";
  }
  if (skip.reason === "busy") {
    const nextMs = schedule.nextRunAt ? Date.parse(schedule.nextRunAt) : Number.NaN;
    if (Number.isFinite(nextMs) && nextMs > now.getTime()) {
      return "上一轮还在跑，已跳到下一次";
    }
  }
  return "";
}

export function overlapPolicyLabel(policy?: string): string {
  return policy === "queue" ? "排队执行" : "上一轮在跑就跳过";
}

export function remainingScheduleQuota(count: number): string {
  const total = Math.max(0, count);
  const left = Math.max(0, MAX_SCHEDULES_PER_BOT - total);
  if (total >= MAX_SCHEDULES_PER_BOT) {
    return `一个 Bot 最多 20 条定时，先删掉不用的再加（当前 ${total}/20）`;
  }
  if (total >= 19) {
    return `还能再加 ${left} 条（${total}/20）`;
  }
  return "";
}

export function apiErrorCode(cause: unknown): string | undefined {
  if (cause && typeof cause === "object" && "code" in cause) {
    const code = (cause as { code?: unknown }).code;
    return typeof code === "string" ? code : undefined;
  }
  return undefined;
}

export function scheduleErrorMessage(
  code: string | undefined,
  fallback: string,
  count?: number,
): string {
  if (code === "SCHEDULE_LIMIT") {
    return remainingScheduleQuota(count ?? MAX_SCHEDULES_PER_BOT);
  }
  if (code === "SCHEDULE_INTERVAL_TOO_SHORT") {
    return "定时间隔最短 5 分钟。太频繁会持续消耗模型额度。";
  }
  return fallback;
}

export function nextEnabledSchedule(schedules: BotSchedule[]): BotSchedule | null {
  const enabled = schedules.filter((item) => item.enabled && item.nextRunAt);
  if (!enabled.length) return null;
  return enabled
    .slice()
    .sort((left, right) => String(left.nextRunAt).localeCompare(String(right.nextRunAt)))[0];
}

export function isScheduleCreateResult(value: unknown): value is BotSchedule & { kind?: string } {
  if (!value || typeof value !== "object") return false;
  const row = value as { kind?: unknown; id?: unknown };
  if (row.kind === "schedule") return true;
  return typeof row.id === "string" && row.id.startsWith("sch_");
}

export function createdScheduleNotice(
  schedule: BotSchedule,
  formatNext: (value?: string | null) => string,
  now: Date = new Date(),
): string {
  const next = schedule.nextRunAt;
  const nextMs = next ? Date.parse(next) : Number.NaN;
  if (schedule.rule?.kind === "once" && Number.isFinite(nextMs) && nextMs <= now.getTime()) {
    return "已设定，时间已过，Pilot 开着时会马上补跑一次。";
  }
  const when = formatNext(next);
  return when ? `已设定，下次 ${when}` : "已设定定时。";
}

export function nextEvening(now: Date): { date: Date; label: string } {
  const date = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 20, 0, 0, 0);
  const tomorrow = date.getTime() <= now.getTime();
  if (tomorrow) date.setDate(date.getDate() + 1);
  return { date, label: `${tomorrow ? "明晚" : "今晚"} 20:00` };
}
